android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageBase.java

```diff
@@ -0,0 +1,489 @@
+  protected void throwIfInsufficientLevel(@NonNull final SecurityLevel level)
+    throws CryptoFailedException {
+
+    if (!securityLevel().satisfiesSafetyThreshold(level)) {
+      throw new CryptoFailedException(String.format(
+        "Insufficient security level (wants %s; got %s)",
+        level, securityLevel()));
+    }
+  }
+
+  /** Extract existing key or generate a new one. In case of problems raise exception. */
+  @NonNull
+  protected Key extractGeneratedKey(@NonNull final String safeAlias,
+                                    @NonNull final SecurityLevel level,
+                                    @NonNull final AtomicInteger retries)
+    throws GeneralSecurityException {
+    Key key;
+
+    do {
+      final KeyStore keyStore = getKeyStoreAndLoad();
+
+      // if key is not available yet, try to generate the strongest possible
+      if (!keyStore.containsAlias(safeAlias)) {
+        generateKeyAndStoreUnderAlias(safeAlias, level);
+      }
+
+      // throw exception if cannot extract key in several retries
+      key = extractKey(keyStore, safeAlias, retries);
+    } while (null == key);
+
+    return key;
+  }
+
+  /** Try to extract key by alias from keystore, in case of 'known android bug' reduce retry counter. */
+  @Nullable
+  protected Key extractKey(@NonNull final KeyStore keyStore,
+                           @NonNull final String safeAlias,
+                           @NonNull final AtomicInteger retry)
+    throws GeneralSecurityException {
+    final Key key;
+
+    // Fix for android.security.KeyStoreException: Invalid key blob
+    // more info: https://stackoverflow.com/questions/36488219/android-security-keystoreexception-invalid-key-blob/36846085#36846085
+    try {
+      key = keyStore.getKey(safeAlias, null);
+    } catch (final UnrecoverableKeyException ex) {
+      // try one more time
+      if (retry.getAndDecrement() > 0) {
+        keyStore.deleteEntry(safeAlias);
+
+        return null;
+      }
+
+      throw ex;
+    }
+
+    // null if the given alias does not exist or does not identify a key-related entry.
+    if (null == key) {
+      throw new KeyStoreAccessException("Empty key extracted!");
+    }
+
+    return key;
+  }
+
+  /** Verify that provided key satisfy minimal needed level. */
+  protected boolean validateKeySecurityLevel(@NonNull final SecurityLevel level,
+                                             @NonNull final Key key)
+    throws GeneralSecurityException {
+
+    return getSecurityLevel(key)
+      .satisfiesSafetyThreshold(level);
+  }
+
+  /** Get the supported level of security for provided Key instance. */
+  @NonNull
+  protected SecurityLevel getSecurityLevel(@NonNull final Key key) throws GeneralSecurityException {
+    final KeyInfo keyInfo = getKeyInfo(key);
+
+    // lower API23 we don't have any hardware support
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      if (keyInfo.isInsideSecureHardware()) {
+        return SECURE_HARDWARE;
```

Comment:
To follow code style used in this file:
```suggestion
        return SecurityLevel.SECURE_HARDWARE;
```